src/context/toastNotificationProvider.tsx

```diff
@@ -0,0 +1,252 @@
+    });
+
+    setCountBySeverity((prev) => {
+      return {
+        ...prev,
+        [notification.type]: (prev[notification.type] || 0) + 1,
+      };
+    });
+
+    showNotificationWithDelay();
+
+    return notificationToAdd;
+  };
+
+  const clear = (notification: ToastNotificationType) => {
+    setNotifications((prev) => {
+      // if we are clearing the last notification from an expanded list,
+      // then we want to collapse the list as well
+      if (prev.length === 1) {
+        setShowList(false);
+      }
+
+      return prev.filter(
+        (existingNotification) => existingNotification.id !== notification.id,
+      );
+    });
+
+    setCountBySeverity((prev) => {
+      return {
+        ...prev,
+        [notification.type]: Math.max(0, (prev[notification.type] || 0) - 1),
+      };
+    });
+
+    clearNotificationTimer();
+  };
+
+  const clearByFilter = (
+    notifications: ToastNotificationType[],
+    filters: Filters,
+  ) => {
+    const newNotifications: ToastNotificationType[] = [];
+    const newCountBySeverity = {
+      caution: 0,
+      information: 0,
+      negative: 0,
+      positive: 0,
+    };
+
+    for (const notification of notifications) {
+      if (!filters[notification.type]) {
+        newNotifications.push(notification);
+        newCountBySeverity[notification.type] += 1;
+      }
+    }
+
+    if (!newNotifications.length) {
+      setShowList(false);
+    }
+
+    setNotifications(newNotifications);
+    setCountBySeverity(newCountBySeverity);
+  };
+
+  const clearAll = () => {
+    setNotifications([]);
+    setCountBySeverity({
+      caution: 0,
+      information: 0,
+      negative: 0,
+      positive: 0,
+    });
+    setShowList(false);
+    clearNotificationTimer();
+  };
+
+  const toggleListView = () => {
+    clearNotificationTimer();
+    setShowList((prev) => !prev);
+  };
+
+  const helper: ToastNotificationHelper = {
+    notifications,
+    failure: (title, error, message, actions) =>
+      addNotification(failure(title, error, message, actions)),
+    info: (message, title) => addNotification(info(message, title)),
+    success: (message) => addNotification(success(message)),
+    clear,
+    clearAll,
+    toggleListView,
+    isListView: showList,
+    countBySeverity,
+  };
+
+  const latestNotification = notifications[notifications.length - 1];
+  const hasNotifications = !!notifications?.length;
+  const showNotification = hasNotifications && !showList && notificationTimer;
+  const showNotificationList = hasNotifications && showList;
+  return (
+    <ToastNotificationContext.Provider value={helper}>
+      {children}
+      <ToastNotification
+        notification={latestNotification}
+        onDismiss={clear}
+        show={!!showNotification}
+      />
+      <ToastNotificationList
+        notifications={notifications}
+        onDismiss={clear}
+        onDismissAll={clearAll}
+        onDismissByFilter={clearByFilter}
```

Comment:
I think we should simplify the interface to a single callback that receives an array of notifications to dismiss. Then we can avoid the three distinct implementations of `clear`, `clearAll` and `clearAllByFilter`. I also think the filters should be a concept of the `ToastNotificationList`, that doesn't leak into the provider.

Reply:
yeah that makes sense, I have combined the 3 methods into `clear` and also moved the filter logic into `ToastNotificationList`

Reply:
nice, thanks :)